google/cloud/bigtable/read_rows_query.py

```diff
@@ -197,18 +240,84 @@ def add_range(
         """
         if not (isinstance(row_range, dict) or isinstance(row_range, RowRange)):
             raise ValueError("row_range must be a RowRange or dict")
-        self.row_ranges.append(row_range)
+        if isinstance(row_range, dict):
+            row_range = RowRange._from_dict(row_range)
+        self.row_ranges.add(row_range)
 
-    def shard(self, shard_keys: "RowKeySamples" | None = None) -> list[ReadRowsQuery]:
+    def shard(self, shard_keys: RowKeySamples) -> list[ReadRowsQuery]:
         """
         Split this query into multiple queries that can be evenly distributed
-        across nodes and be run in parallel
+        across nodes and run in parallel
 
         Returns:
             - a list of queries that represent a sharded version of the original
               query (if possible)
+        Raises:
+            - AttributeError if the query contains a limit
         """
-        raise NotImplementedError
+        if self.limit is not None:
+            raise AttributeError("Cannot shard query with a limit")
+        if len(self.row_keys) == 0 and len(self.row_ranges) == 0:
+            # empty query represents full scan
+            # ensure that we have at least one key or range
+            full_scan_query = ReadRowsQuery(
+                row_ranges=RowRange(), row_filter=self.filter
+            )
+            return full_scan_query.shard(shard_keys)
+
+        split_points = [sample[0] for sample in shard_keys if sample[0]]
+        sharded_queries: dict[int, ReadRowsQuery] = defaultdict(
+            lambda: ReadRowsQuery(row_filter=self.filter)
+        )
+
+        # use binary search to find split point segments for each row key in original query
+        # each split point represents the inclusive end key for that segment
+        for this_key in list(self.row_keys):
+            segment_index = bisect.bisect_left(split_points, this_key)
+            sharded_queries[segment_index].add_key(this_key)
+
+        # use binary search to find start and end segments for each row range in original query
+        # if range spans multiple segments, split it into multiple ranges
+        for this_range in self.row_ranges:
+            # start either bisects left or right, depending on whether the range start is inclusive
+            if this_range.start is None:
+                start_segment = 0
+            elif this_range.start.is_inclusive:
+                start_segment = bisect.bisect_left(split_points, this_range.start.key)
+            else:
+                start_segment = bisect.bisect_right(split_points, this_range.start.key)
+            # split points represent end of range; always bisect left
+            end_segment = (
+                bisect.bisect_left(split_points, this_range.end.key)
+                if this_range.end is not None
+                else len(split_points)
+            )
```

Comment:
This is very hard to read. It looks like bisect_left results are ignored. I think using statement form of if would make it a lot easier to read:

```
if this_range.end is not None:
   end_segment = bisect.bisect_left(split_points, this_range.end.key)
else:
   end_segment = len(split_points)
```

Also there is a small opportunity for an optimization of starting the binary search from the start_segment